packages/eslint-plugin/src/rules/no-unnecessary-type-parameters.ts

```diff
@@ -108,7 +108,21 @@ export default createRule({
                 for (const reference of smTypeParameterVariable.references) {
                   if (reference.isTypeReference) {
                     const referenceNode = reference.identifier;
-                    yield fixer.replaceText(referenceNode, constraintText);
+                    const isCompositeType =
+                      constraint?.type === AST_NODE_TYPES.TSUnionType ||
+                      constraint?.type === AST_NODE_TYPES.TSIntersectionType;
+                    const hasMatchingAncestorType = [
+                      AST_NODE_TYPES.TSArrayType,
+                      AST_NODE_TYPES.TSIndexedAccessType,
+                    ].some(type => referenceNode.parent.parent?.type === type);
```

Comment:
There's a missing code coverage point here, as reported by Codecov. Removing the `?` doesn't fail the test. 

It looks to me like another example of #6255 / #10682 where the AST types for `parent` nodes aren't precise enough to know the grandparent of an identifier is defined. If that's the case we'll want a `!` type assertion and a followup issue. Or, if not, there's a missing unit test for a case that could be hit. Could you please take a look?

Reply:
Based on my investigation, it seems likely that when `reference.isTypeReference` is `true`, the `type` of `reference.identifier.parent` is `TSESTree.TSTypeReference`.
I agree that a follow-up might be necessary, as you suggested.
I have added the ! type assertion as well. → [91dd50d](https://github.com/typescript-eslint/typescript-eslint/pull/10907/commits/91dd50d9df0d4758deba94405cba78a6e2b4459e)

This is just a quick idea, but it seems like we might need some form of type narrowing:
```ts
  public isTypeReference(): this is Reference & {
    identifier: (TSESTree.Identifier | TSESTree.JSXIdentifier) & {
      parent: TSESTree.TSTypeReference;
    };
  } {
    return (this.#referenceType & ReferenceTypeFlag.Type) !== 0;
  }
```

Reply:
Oh joy, a refactor. Yes, an issue would be good. I suspect we don't have much freedom in changing the shape of `isTypeReference`.